extensions/amp-lightbox/1.0/base-element.js

```diff
@@ -47,6 +47,16 @@ export class BaseElement extends PreactBaseElement {
     toggle(this.element, this.open_);
     this.triggerEvent(this.element, this.open_ ? 'open' : 'close');
   }
+
+  /** @override */
+  mutationObserverCallback() {
+    const open = this.element.hasAttribute('open');
+    if (open === this.open_) {
+      return;
+    }
+    this.open_ = open;
+    open ? this.api().open() : this.api().close();
```

Comment:
So we're always closing when we don't have `[open]`? Is it always set when open?

I'm also interested in understanding why we have two ways to interpret whether we're open: the local property and the element attribute.

Reply:
> So we're always closing when we don't have [open]?
Yes

> Is it always set when open?
Yes: 
https://github.com/ampproject/amphtml/blob/5c8cee4e06e5d171b241a6cacb16e88b5260616b/extensions/amp-lightbox/1.0/base-element.js#L33-L49

> I'm also interested in understanding why we have two ways to interpret whether we're open: the local property and the element attribute.
The local property is to track if the `open` attribute actually was the one that changed to trigger the `mutationObserverCallback` - i.e. if something else mutated on the element to cause this callback to fire, we don't want to do anything. The reason we don't specify/watch the attribute `open` in `BaseElement['props']` or pass `open` into the component is because of the two-step animation sequence (mount and then transition visible), which is better triggered via API than declaratively. 

But really this change was moving code that already existed to a different file, looks like it was missed in the restructuring process. The original contribution of this behavior was in #31544 for context.

Reply:
FMI: What is the benefit of moving this to BaseElement?  Or is this a style / convention thing?

Reply:
This was moved to `BaseElement` because we want this behavior for custom elements on non-AMP pages as well as AMP pages. Its prior location would have only provided the behavior to AMP pages.

Reply:
I see, thanks, that makes sense!